src/Microsoft.Health.Dicom.Functions/Indexing/ReindexOperation.Activity.cs

```diff
@@ -0,0 +1,136 @@
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Extensions.DurableTask;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Dicom.Core.Features.ExtendedQueryTag;
+using Microsoft.Health.Dicom.Core.Features.Indexing;
+using Microsoft.Health.Dicom.Functions.Indexing.Models;
+
+namespace Microsoft.Health.Dicom.Functions.Indexing
+{
+    /// <summary>
+    /// Represents the Azure Durable Functions that perform the re-indexing of previously added DICOM instances
+    /// based on new tags configured by the user.
+    /// </summary>
+    public partial class ReindexOperation
+    {
+        /// <summary>
+        /// The activity to complete operation.
+        /// </summary>
+        /// <param name="operationId">The operation id.</param>
+        /// <param name="log">The log.</param>
+        /// <returns>The task.</returns>
+        [FunctionName(nameof(CompleteOperationAsync))]
+        public Task CompleteOperationAsync([ActivityTrigger] string operationId, ILogger log)
+        {
+            EnsureArg.IsNotNull(log, nameof(log));
+
+            log.LogInformation("Completing Reindex operation on {operationId}", operationId);
+            return _tagOperationStore.CompleteOperationAsync(operationId);
+        }
+
+        /// <summary>
+        /// The activity to start reindex operation.
+        /// </summary>
+        /// <param name="input">The input.</param>
+        /// <param name="log">The log.</param>
+        /// <returns>The task.</returns>
+        [FunctionName(nameof(StartOperationAsync))]
+        public Task StartOperationAsync([ActivityTrigger] StartOperationInput input, ILogger log)
+        {
+            EnsureArg.IsNotNull(log, nameof(log));
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            log.LogInformation("Starting reindex operation with input {input}", input);
+            return _tagOperationStore.StartOperationAsync(input.OperationId, input.TagStoreEntries);
+        }
+
+        /// <summary>
+        ///  The activity to add extended query tags.
+        /// </summary>
+        /// <param name="entries">The entries.</param>
+        /// <param name="log">The log.</param>
+        /// <returns>The store entries.</returns>
+        [FunctionName(nameof(AddExtendedQueryTagAsync))]
+        public async Task<IEnumerable<ExtendedQueryTagStoreEntry>> AddExtendedQueryTagAsync([ActivityTrigger] IEnumerable<AddExtendedQueryTagEntry> entries, ILogger log)
+        {
+            EnsureArg.IsNotNull(entries, nameof(entries));
+            EnsureArg.IsNotNull(log, nameof(log));
+            log.LogInformation("Adding extended query tags {entries}", entries);
+            var response = await _addExtendedQueryTagService.AddExtendedQueryTagAsync(entries);
```

Comment:
This is just echoing the input?

Reply:
sorry, not get you. It just pass to input to AddExtendedQueryTagService which previously proceed the Add.

Reply:
Yeah, so the method just returns its input. Eg. `response` is the same as `entries`?

Reply:
* After query tag saved to database, a tag key is appended, the return should include that.
* If any of tags fails to add (e.g: already exist), the whole add fails.

Reply:
this should have been resolved.